internal/app/machined/pkg/controllers/siderolink/manager.go

```diff
@@ -110,80 +107,173 @@ func parseAPIEndpoint(sideroLinkParam string) (apiEndpoint, error) {
+	wgClient, wgClientErr := wgctrl.New()
+	if wgClientErr != nil {
+		return wgClientErr
 	}
 
-	conn, err := grpc.DialContext(ctx, parsedEndpoint.Host, grpc.WithTransportCredentials(transportCredentials))
-	if err != nil {
-		return fmt.Errorf("error dialing SideroLink endpoint %q: %w", apiEndpoint, err)
+	defer func() {
+		if closeErr := wgClient.Close(); closeErr != nil {
+			logger.Error("failed to close wg client", zap.Error(closeErr))
+		}
+	}()
+
+	var zeroKey wgtypes.Key
+
+	if bytes.Equal(ctrl.nodeKey[:], zeroKey[:]) {
+		var err error
+
+		ctrl.nodeKey, err = wgtypes.GeneratePrivateKey()
+		if err != nil {
+			return fmt.Errorf("error generating Wireguard key: %w", err)
+		}
 	}
 
-	sideroLinkClient := pb.NewProvisionServiceClient(conn)
+	ticker := time.NewTicker(30 * time.Second)
+	defer ticker.Stop()
 
 	for {
 		select {
 		case <-ctx.Done():
 			return nil
+		case <-ticker.C:
+			reconnect, err := ctrl.shouldReconnect(wgClient)
+			if err != nil {
+				return err
+			}
+
+			if !reconnect {
+				// nothing to do
+				continue
+			}
 		case <-r.EventCh():
 		}
 
-		netStatus, err := r.Get(ctx, resource.NewMetadata(network.NamespaceName, network.StatusType, network.StatusID, resource.VersionUndefined))
+		linkSpec := network.NewLinkSpec(network.ConfigNamespaceName, network.LayeredID(network.ConfigOperator, network.LinkID(constants.SideroLinkName)))
+
+		cfg, err := safe.ReaderGet[*siderolink.Config](ctx, r, siderolink.NewConfig(config.NamespaceName, siderolink.ConfigID).Metadata())
 		if err != nil {
 			if state.IsNotFoundError(err) {
-				// no network state yet
+				if cleanupErr := ctrl.cleanup(ctx, r, linkSpec.Metadata()); cleanupErr != nil {
+					return fmt.Errorf("failed to do cleanup: %w", cleanupErr)
+				}
+
+				// no config
 				continue
 			}
 
-			return fmt.Errorf("error reading network status: %w", err)
+			return fmt.Errorf("failed to get siderolink config: %w", err)
 		}
 
-		if !netStatus.(*network.Status).TypedSpec().AddressReady {
-			// wait for address
-			continue
+		sysInfo, err := safe.ReaderGet[*hardware.SystemInformation](ctx, r, hardware.NewSystemInformation(hardware.SystemInformationID).Metadata())
+		if err != nil {
+			if state.IsNotFoundError(err) {
+				// no system information
+				continue
+			}
+
+			return fmt.Errorf("failed to get system information: %w", err)
+		}
+
+		nodeUUID := sysInfo.TypedSpec().UUID
+		endpoint := cfg.TypedSpec().APIEndpoint
+
+		parsedEndpoint, err := parseAPIEndpoint(endpoint)
+		if err != nil {
+			return fmt.Errorf("failed to parse siderolink endpoint: %w", err)
+		}
+
+		var transportCredentials credentials.TransportCredentials
+
+		if parsedEndpoint.Insecure {
+			transportCredentials = insecure.NewCredentials()
+		} else {
+			transportCredentials = credentials.NewTLS(&tls.Config{})
+		}
+
+		provision := func() (*pb.ProvisionResponse, error) {
+			conn, connErr := grpc.DialContext(ctx, parsedEndpoint.Host, grpc.WithTransportCredentials(transportCredentials))
```

Comment:
I would probably set here some timeout on the context, just to make sure we don't hang for too long, and better retry

Reply:
done